### DISCUSSION TITLE
Possible type narrowing bug

### DISCUSSION THREAD

**Discussion by tpgillam** [microsoft/pyright#8948] [General]
Tested on pyright 1.1.379, python 3.11.9

```python
import typing

T = typing.TypeVar("T", int, str)

def moo(x: T , y: T):
    if isinstance(x, int):
        typing.assert_type(y, int)  # "assert_type" mismatch: expected "int" but received "T@moo" [reportAssertTypeFailure]
```

Intuitively it seems like `y` should be narrowed here. And also in the looser case where `T` isn't constrained. I might be missing something though?

**Comment by erictraut** [comment]
Pyright is working as intended here. Type narrowing applies only to subexpressions that are found in the type guard expression (the conditional check). It does not track relationships between types of symbols.
